### PR TITLE
Add .git to builds

### PR DESCRIPTION
**Pull request by ianblenke**
feat(deis-builder/add_dot_git): Add .git to builds

The current deis git push behavior strips .git from the build. This happens during the "git archive" step in the deis-builder template/builder script. In order to use submodules, the "git archive" method of sending the tarball to "docker build" needs to be replaced. As the repo is bare, it needs to be cloned to the TMP_DIR and checked out to the $GIT_SHA of the deployment. This also allows for the use of submodules in deployments.

### CODE REVIEW

File: builder/image/templates/builder [hunk 1]
```diff
@@ -56,7 +56,12 @@ mkdir -p $BUILD_DIR $CACHE_DIR
 TMP_DIR=$(mktemp -d --tmpdir=$BUILD_DIR)
 
 cd $REPO_DIR
-git archive $GIT_SHA | tar -xmC $TMP_DIR
+git clone $REPO_DIR/ "$TMP_DIR/"
+pushd "$TMP_DIR"
```
Comment (bacongobbler) [inline builder/image/templates/builder]: on line 67 we `cd` to $TMP_DIR. It'd be cleaner if we moved these operations below that line and removed the call to `cd $REPO_DIR` on line 58.

File: builder/image/templates/builder [hunk 2]
```diff
@@ -56,7 +56,12 @@ mkdir -p $BUILD_DIR $CACHE_DIR
...
+export GIT_DIR="$TMP_DIR/.git" GIT_WORK_TREE="$TMP_DIR"
+git config core.bare false
+git reset --hard $GIT_SHA
```
Comment (bacongobbler) [inline builder/image/templates/builder]: why not a `git checkout` instead? That way we're not destroying the tree's history

**Comment by ianblenke** [comment]
Thankfully, in this context, we aren't destroying anything.

The TMP_DIR is actually cloned from a bare git repo, it is empty except for the .git directory that was copied over.

That TMP_DIR is indeed temporary (see the mktemp a few lines up). As such, there is no history there to be lost.

The big problem here is that you can't simply do a "git checkout $GIT_SHA" as it will blow up telling you that the .git/HEAD mentions "refs/origin/master" and that doesn't exist if you didn't git push a master branch yet. So, what you end up doing, is doing a: "git checkout $(git branch -a)" followed by a "git checkout $GIT_SHA", all to end up with effectively the same thing you would have had by doing a "git reset --hard $GIT_SHA" in the first place. One way you would be in a local branch named after whatever branch they pushed checked out to the specific SHA1, the other way you end up in a disassociated branch checked out to the specific SHA1. The net effect is ambiguous after build time anyway, it's really a question of what the slug runtime sees when it tries to use that .git tree.

**Comment by bacongobbler** [comment]
After a bit of reading between the two, essentially they do the same thing for this operation so I guess it doesn't really matter either way.

**Comment by carmstrong** [comment]
@ianblenke would this close #1094?

**Comment by carmstrong** [comment]
Code LGTM.

**Comment by mboersma** [comment]
Code LGTM. We should test this against a project with submodules.

**Comment by ianblenke** [comment]
Aye, this would potentially close #1094, with the caveat that you still must manage your ssh keys with your own buildpack (which we might build an example of) or in a Dockerfile as a RUN step. This fix merely makes .git available for use by a buildpack/Dockerfile.

In some cases, it might actually impair customers who expect a smaller resultant layer due to the fact that their .git has been ignored until now.

It might not be a bad idea to make this behavior conditional somehow to avoid that.

Principle of least surprise though: which is more surprising, finding that .git isn't included when doing a deploy, or finding that your smallish images are suddenly larger thanks to .git being included in a newer version of Deis?

**Comment by bacongobbler** [comment]
IIRC Heroku strips this directory from the final slug. It might be better to clone the repository, initialize submodules then strip .git so the behaviour stays the same.

**Comment by ianblenke** [comment]
The question then becomes when to initialize the submodules. If we move that functionality into the deis build script, we have to take into account private credentials for private git repos. If we leave it up to the buildpack/Dockerfile, do we leave it up to those mechanisms to make that decision or whether or not to strip the .git folder? While we could conceivably strip the .git folder after building a slug using a buildpack, we can't really remove the space overhead in a layer generated by an app's Dockerfile - unless we just want to flatten all such app Dockerfile generates images into a slug-like form before deploying to further help _possibly_ minimizing deploys (but not sharing anything between deploys as a result)...

**Comment by bacongobbler** [comment]
hmm, I'm wondering if we should close this in favour of #3259 since that seems to cover all bases proposed here (private buildpacks, SSH keys, **and** heroku submodules).

**Comment by ianblenke** [comment]
@bacongobbler, I ran into problems attempting to use a shallow git clone with submodules, so I don't believe we can use the `--depth 1` option as in #3259.

**Comment by bacongobbler** [comment]
> I ran into problems attempting to use a shallow git clone with submodules, so I don't believe we can use the --depth 1 option as in #3259

I don't see a `--depth=1` being used in #3259..?

**Comment by ianblenke** [comment]
Unfortunately #3305 has to do with ssh keys. So does #3259. While that does help access to private repos, and is an important thing to consider as it is tangentially related to this issue, neither really addresses the submodules _versioning_ problem.

According to this StackOverflow post, the commit sha1 of the submodule "[is stored in Git's object database directly. The tree object for the directory where the submodule lives will have an entry for the submodule's commit (this is the so-called "gitlink")](http://stackoverflow.com/questions/5033441/where-does-git-store-the-sha1-of-the-commit-for-a-submodule)".

I just tried to create a pair of local git repos and have one file:// reference the other as a submodule, and I could shallow clone that with `--depth=1` and it would indeed retain the submodule sha1, even after having added a few commits. The submodule sha1 metadata is being retained even in a shallow clone.

The point of this PR is really to make sure we clone the entire tree, including the .git folder. Without .git, there is no way to access this sha1 version metadata, and that is the larger problem that the ssh related issues don't address.

**Comment by withinboredom** [comment]
@ianblenke If you are looking for a git repo with lots of submodules and commits, see https://github.com/FuturesJS/FuturesJS - it isn't an app, but you may find it useful to test with what you're working on.

**Comment by deis-admin** [comment]
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

**Comment by bacongobbler** [comment]
@ianblenke after our PR review this morning, we evaluated this and felt like this is a very special use case. This is a perfect use case for customizing the builder, but considering that the SSH_KEY PR should cover most of problems towards initializing submodules we're going to close this as a wontfix.

**Pull request closed without merging**